omp/reorder/rcm_kernels.cpp

```diff
@@ -0,0 +1,904 @@
+    std::shared_ptr<const OmpExecutor> exec, const size_type num_vertices,
+    const IndexType *const row_ptrs, const IndexType *const col_idxs,
+    const IndexType *const degrees,
+    IndexType *const levels,  // Must be max/inf in all nodes connected to start
+    const IndexType start, const IndexType max_degree)
+{
+    // Create a level structure.
+    ubfs(exec, num_vertices, row_ptrs, col_idxs, levels, start, max_degree);
+
+    // Now find the node in the last level with minimal degree.
+    // TODO: Rewrite this as a custom reduction.
+
+    // First local ...
+    const auto num_threads = omp_get_max_threads();
+    vector<IndexType> local_contenders(num_threads, exec);
+    vector<IndexType> local_degrees(num_threads, exec);
+    vector<IndexType> local_heights(num_threads, exec);
+
+#pragma omp parallel num_threads(num_threads)
+    {
+        const auto tid = omp_get_thread_num();
+        auto local_contender = 0;
+        auto local_degree = std::numeric_limits<IndexType>::max();
+        auto local_height = 0;
+
+#pragma omp for schedule(static)
+        for (auto i = 0; i < num_vertices; ++i) {
+            if (levels[i] > local_height) {
+                local_contender = i;
+                local_degree = degrees[i];
+                local_height = levels[i];
+            } else if (levels[i] == local_height && degrees[i] > local_degree) {
+                local_contender = i;
+                local_degree = degrees[i];
+            }
+        }
+
+        local_contenders[tid] = local_contender;
+        local_degrees[tid] = local_degree;
+        local_heights[tid] = local_height;
+    }
+
+    // ... then global.
+    auto global_contender = local_contenders[0];
+    auto global_degree = local_degrees[0];
+    auto global_height = local_heights[0];
+    for (auto i = 1; i < num_threads; ++i) {
+        if (local_heights[i] > global_height) {
+            global_contender = local_contenders[i];
+            global_degree = local_degrees[i];
+            global_height = local_heights[i];
+        } else if (local_heights[i] == global_height &&
+                   local_degrees[i] > global_degree) {
+            global_contender = local_contenders[i];
+            global_degree = local_degrees[i];
+        }
+    }
+
+    return std::make_pair(global_contender, global_height);
+}
+
+
+/**
+ * Finds the index of a node with minimum degree and the maximum degree.
+ */
+template <typename IndexType>
+std::pair<IndexType, IndexType> find_min_idx_and_max_val(
+    std::shared_ptr<const OmpExecutor> exec, const size_type num_vertices,
+    const IndexType *const row_ptrs, const IndexType *const col_idxs,
+    const IndexType *const degrees, vector<IndexType> &levels,
+    const uint8 *const previous_component,
+    gko::reorder::starting_strategy strategy)
+{
+    // First find local extrema.
+    const auto num_threads = omp_get_max_threads();
+    vector<std::pair<IndexType, IndexType>> local_min_vals_idxs(num_threads,
```

Comment:
Ah, there the critical error that I already tracked down once, but subsequently forgot to fix, is again.
"Index" here means "node (identifier)". How many nodes can there be? Well, `num_vertices` is `size_type`, **but** we write our result into a permutation, that is `IndexType` - meaning there can only be `IndexType` nodes.
The original interface for `get_permutation` is incoherent. I suggest changing `size_type` to `IndexType` for the number of vertices.

Reply:
That sounds good to me, I will mark this as resolved then.
We actually plan to change `size_type` from `uint64` to `int64` to prevent these annoying warnings of using `unsigned` with `signed` types together. However, since that is a interface change, we can only do it in `Ginkgo 2.0`, which can take a while.